Chunk Data Model supports per-chunk service event mapping

jordanschalm opened the pull request:
This PR adds support for specifying which service events were emitted in which chunk, by modifying the `ChunkBody` data structure in a backward compatible manner. Addresses #6622.

### Changes
- Adds `ServiceEventCount` field to `ChunkBody`:
  - This field creates an explicit mapping, committed to by Execution Nodes, of service event to chunk. This allows Verification Nodes to know which service events to expect when validating any chunk.
  - This field is defined to be backward-compatible with prior data model versions:
    - Existing serializations of `ChunkBody` will unmarshal into a struct with a nil `ServiceEventCount`. We define a chunk with nil `ServiceEventCount` to have the same semantics as before the field existed: if any service events were emitted, then they were emitted from the system chunk.
    - Post software upgrade, all new (honest) serializations of `ChunkBody` will always have non-nil `ServiceEventCount`

### Upgrade Notes
Source of truth for upgrade plans (still WIP): https://flowfoundation.notion.site/EFM-Recovery-Release-Upgrade-Plan-WIP-14d1aee1232480228a87e43933815285?pvs=4

**Note:** Implementation changes associated with the upgrade process will be implemented separately, when the upgrade process is fully specified (see https://github.com/onflow/flow-go/issues/6777).

https://github.com/onflow/flow-go/issues/6783 captures changes required around the upgrade behaviour.

### To Do Before Merging
- [x] Ensure consistent hashing between `ExecutionResult` versions
- [x] Update description in https://github.com/onflow/flow-go/issues/6773
- [x] Check if necessary to update RPC model (rpc conversion tests fail with non-nil `ServiceEventCount` field)
- [x] Should this directly target master / the current spork branch, as this HCU will occur before EFM Recovery? (See [upgrade plan](https://www.notion.so/flowfoundation/EFM-Recovery-Release-Upgrade-Plan-14d1aee1232480228a87e43933815285?pvs=4))

---

This PR replaces two prior approaches, implemented in part https://github.com/onflow/flow-go/pull/6629 and https://github.com/onflow/flow-go/pull/6730.

AlexHentschel commented:
## strategic / conceptual thoughts
1. I am thinking about enforcing activation of the protocol extension (specifically, the usage of the [`Chunk.ServiceEventIndices` field](https://github.com/onflow/flow-go/blob/718a5eee0957aa513bac134f095e1426b5a2cb35/model/flow/chunk.go#L45-L52)). For example, it would be nice if consensus nodes could drop Execution Results with the deprecated format and avoid incorporating them into a block (and rejecting blocks that still do). 
2. To ensure consensus incorporates only execution receipts following the new convention after a certain height, it would be great if we could include some consistency check also in the [`receiptValidator` (somewhere around here)](https://github.com/onflow/flow-go/blob/718a5eee0957aa513bac134f095e1426b5a2cb35/module/validation/receipt_validator.go#L72). 
3. I was wondering if your plan is still to remove [`Chunk.ServiceEventIndices` field](https://github.com/onflow/flow-go/blob/718a5eee0957aa513bac134f095e1426b5a2cb35/model/flow/chunk.go#L45-L52) for Byzantine Fault Tolerance in the long term? I  think we had talked about turning `ExecutionResult.ServiceEventList`  into an indexed list. Or are you thinking about keeping the [`ServiceEventIndices` field](https://github.com/onflow/flow-go/blob/718a5eee0957aa513bac134f095e1426b5a2cb35/model/flow/chunk.go#L45-L52) as the long-term solution --  just with removing the backwards-compatibility case?

   In general, my preference would be to also allow `Chunk.ServiceEventIndices = nil` when there are no service events generated in the chunk. Thereby the final solution becomes a lot more intuitive: 
   * if [`ExecutionResult.ServiceEvents`](https://github.com/onflow/flow-go/blob/718a5eee0957aa513bac134f095e1426b5a2cb35/model/flow/execution_result.go#L16) is not empty (nil allowed), then the _new_ convention requires for consistency:

       $$\sum_\texttt{Chunk} \texttt{len}(Chunk.ServiceEventIndices) = \texttt{len}(\texttt{ExecutionResult.ServiceEventList})\qquad\qquad\qquad\qquad(1) $$ 

      I feel is check is very similar to other stuff, which consensus nodes already verify about an `Execution Receipt` (see [`ReceiptValidator` implementation](https://github.com/onflow/flow-go/blob/718a5eee0957aa513bac134f095e1426b5a2cb35/module/validation/receipt_validator.go#L72))
   * We would temporary relax the new convention, eq. $(1)$,  for downwards compatibility as follows: the `ServiceEventIndices` fields of _all_ chunks can be nil despite there being service events. 
   * As service events are rare, `ExecutionResult.ServiceEventList` is empty in the majority cases. Then both the deprecated as well as the new conventions would allow `ChunkBody.ServiceEventIndices` to be `nil` (which is the most intuitive convention anyway). Also, for individual chunks that don't produce any service events, their `ChunkBody.ServiceEventIndices` could be `nil` or empty according to the new convention. Then, also the new convention is very self-consistent in my opinion - and the depreciation condition is only an add on that can be later removed. 



I mean is this is only a temporary solution, I am happy and we can skip most of question 3.

jordanschalm commented:
## Summary of Discussion with @AlexHentschel 

### Change of `ServiceEventIndices` structure
- Replaces list with `ServiceEventsNumber`, a uint16 which is a count of the number of service events emitted in that chunk
  - Since VNs have access to all chunks, they can easily compute the index range based on this field alone
  - This is a more compact, and fixed-size representation
  - Structural validation is simpler, we require only that `sum(chunk.ServiceEventsNumber for chunk in chunks) == len(ServiceEvents)`
- Backward compatibility:
  - If any service events are emitted in a result, and all `ServiceEventsNumber` fields are 0, then this is interpreted as a v0 model version: all service events must have been emitted in the system chunk.

### Removing backward-compatibility at next spork
We plan to keep the overall structure as the long term solution, and only remove the backward-compatibility support at the next spork. We do not plan to use a different representation (ie. [Chunk.ServiceEventIndices field](https://github.com/onflow/flow-go/blob/718a5eee0957aa513bac134f095e1426b5a2cb35/model/flow/chunk.go#L45-L52)).

### Upgrade Comments
- Protocol HCUs are triggered by a `ProtocolVersionUpgade` service event. This service event is emitted outside the system chunk, meaning that **the first Protocol HCU must take place after the service event validation fix has been deployed.**
- We plan to incorporate all changes under `feature/efm-recovery` in one Protocol HCU.

#### Rough Outline of Process
1. Do a manual rolling upgrade of all node roles, to a version including `feature/efm-recovery`.
2. Manually verify that all ENs are upgraded.
  - this is necessary prior to emitting the first ex-system-chunk service event
  - ideally, VNs are also updated, but we can rely on emergency sealing as a fallback if necessary
3. Emit ProtocolVersionUpgrade service event, scheduling the protocol upgrade at view V.
  - Nodes which are not upgraded when we enter view V will halt.
  - We must have a substantial majority (>> supermajority) of SNs and LNs for the network to remain live, before entering view V

jordanschalm commented:
## Summary of discussion with @zhangchiqing 

Let's call v0 the old version and v1 the new version. Leo pointed out the v0 nodes will fail to validate data models produced by v1 nodes. In particular, this step:
> Do a manual rolling upgrade of all node roles, to a version including feature/efm-recovery.

will not work, if v1 nodes immediately begin produce v1 chunks with non-nil `ServiceEventsNumber` field. 

This means the upgrade needs to be split into multiple steps, because v0 software will be unable to read v1 data models with non-nil new fields (will produce different hashes). Instead, we need:

1.  Rolling upgrade from software version v0 to v1 (v1 still produces chunks with nil field)
2.  Protocol HCU, after which v0 nodes cannot progress. At this point, v1 nodes begin producing v1 chunk models with non-nil fields. Chunk models with nil fields are also still accepted (due to sealing lag).
3. Protocol HCU, after which v0 data models are no longer accepted in new blocks

Because of the above additional complexity, I think we should revert the `ProtocolStateVersionUpgrade` service event to be emitted in the system chunk (at least to start), because we need Protocol HCUs to safely roll out the breaking chunk model change. After the upgrade is complete, all subsequent service events may be emitted outside the system chunk.

zhangchiqing commented:
I think we can still do with just one HCU, but we might need 2 rolling upgrades. This is the steps, basically Step 1 and Step 3 are rolling upgrades, Step 2 is an protocol HCU:

1. Rolling upgrade from software version v0 to v1 (v1 still produces chunks with nil field)
Making sure that:
- v1 EN will produce v1 result that have chunks with nil field, so that v1 result can be accepted by both v0 SN and v1 SN.
- v1 result with nil field in chunk produce the same result ID as the v0 result decoded from the v1 result.
- v1 block with the v1 result that has nil field, also produce the same block ID as the v0 block with the decoded v0 result, so that v1 block can be accepted by both v0 nodes or v1 nodes.

2. Protocol HCU, after which results with nil field will not be accepted. At this point, v1 ENs begin producing v1 chunk models with non-nil fields for blocks above HCU height. Details:

Once an protocol HCU event is emitted, then:
- When v1 SN receiving results, it will reject results with nil field below the HCU height.
- When v1 SN building blocks for height above the HCU height, it will include v1 results with non-nil fields only. Note, due to sealing lag, v1 SN might build some blocks that still have some results with nil field, because they are for blocks below the HCU height. 
- When any v1 nodes receiving blocks, it will reject any block that contains result above the HCU height that has nil field in chunks
- Blocks produced from v0 SN for blocks above HCU won’t be accepted by v1 SN
- v0 SN cannot process after HCU because it won’t accept blocks from v1 SN (the block hash doesn’t match)


3. Rolling upgrade that v0 result or v1 result with nil field will not be accepted in new blocks
- Once HCU height in step 2 has been sealed, and all ENs are on v1, we can roll out this upgrade, because we are sure all results in new blocks won’t have nil field in chunks.

jordanschalm commented:
@zhangchiqing I agree that would work. Though, I suspect we will prefer fewer _software upgrades_, so we depend less on actions by node operators, but it's good to know that we have flexibility.

I have written a version of the upgrade plan based on our last discussion (with 2 HCUs and 1 rolling upgrade) [here](https://www.notion.so/flowfoundation/EFM-Recovery-Release-Upgrade-Plan-WIP-14d1aee1232480228a87e43933815285?pvs=4#1511aee1232480dd8972c5050764ca57). The upgrade process is fairly simple - there is also an enumeration of versions, which I hope can be generalized beyond this specific example.

AlexHentschel commented on module/chunks/chunkVerifier_test.go:
```diff
@@ -246,7 +252,7 @@ func (s *ChunkVerifierTestSuite) TestEventsMismatch() {
 // TestServiceEventsMismatch tests verification behavior in case
 // of emitted service events not matching chunks'
 func (s *ChunkVerifierTestSuite) TestServiceEventsMismatch() {
```
For clarity, I would suggest to rename a couple of the existing tests, to make the conceptual difference more explicit:
```suggestion
func (s *ChunkVerifierTestSuite) TestServiceEventsMismatch_SystemChunk() {
```

AlexHentschel commented on module/chunks/chunkVerifier_test.go:
```diff
@@ -269,7 +276,7 @@ func (s *ChunkVerifierTestSuite) TestServiceEventsMismatch() {
 
 // TestServiceEventsAreChecked ensures that service events are in fact checked
 func (s *ChunkVerifierTestSuite) TestServiceEventsAreChecked() {
```
For clarity, I would suggest to rename a couple of the existing tests, to make the conceptual difference more explicit:
```suggestion
func (s *ChunkVerifierTestSuite) TestServiceEventsAreChecked_SystemChunk() {
```

AlexHentschel reviewed (approved):
Looks great. I struggled with understanding the [Chunk Verifier Tests](https://github.com/onflow/flow-go/blob/176100facfc9a0f8f3b4a1daf47bdd8f5de4043d/module/chunks/chunkVerifier_test.go#L268). I think we can fix this with some documentation. They are tests, its fine if the documentation is repetitive.

jordanschalm commented on model/flow/chunk.go:
```diff
@@ -20,19 +22,112 @@ func init() {
+	// (2) Otherwise, ServiceEventCount must be non-nil.
+	// Within an ExecutionResult, all chunks must use either representation (1) or (2), not both.
+	ServiceEventCount *uint16
+	BlockID           Identifier // Block id of the execution result this chunk belongs to
```
I agree, but RLP encoding depends on field ordering within structs, so doing this would change the ID computation (unless we over-rode again, using the RLP encoding).

jordanschalm replied:
Added a test case to validate this https://github.com/onflow/flow-go/pull/6744/commits/a61e30ed7a5aed77038e1f40a97a86a5289d5008

jordanschalm commented on model/flow/chunk.go:
```diff
@@ -20,19 +22,112 @@ func init() {
+//   - when computing the ID of such a ChunkBody, the ServiceEventCount field is omitted from the fingerprint
+func (ch ChunkBody) EncodeRLP(w io.Writer) error {
+	var err error
+	if ch.ServiceEventCount == nil {
```
I think we can remove the `EncodeRLP` method entirely instead. The other branch is essentially the default RLP behaviour.

jordanschalm commented on utils/unittest/encoding.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+// CopyStructure copies public fields of compatible types from src to dst.
+func CopyStructure(t *testing.T, src, dst any) {
```
https://github.com/onflow/flow-go/pull/6744/commits/6fe51f6a5e3d37f7752219820547c94798c8390d

jordanschalm commented on model/flow/execution_result_test.go:
```diff
@@ -41,3 +43,137 @@ func TestExecutionResultGroupBy(t *testing.T) {
+
+	t.Run("service events only in system chunk", func(t *testing.T) {
+		t.Run("nil ServiceEventCount field (old model)", func(t *testing.T) {
+			result := unittest.ExecutionResultFixture()
```
🦅 Good eye - fixed in https://github.com/onflow/flow-go/pull/6744/commits/4a27b73a81734af7a446945b5591a0b617de1d3a

zhangchiqing reviewed (approved):
Nice tests. LGTM

jordanschalm commented on module/chunks/chunkVerifier_test.go:
```diff
@@ -282,9 +289,54 @@ func (s *ChunkVerifierTestSuite) TestServiceEventsAreChecked() {
+	script := "service event mismatch in non-system chunk"
+	meta := s.GetTestSetup(s.T(), script, false, true)
+	vch := meta.RefreshChunkData(s.T())
+
+	// modify the list of service events produced by FVM
+	// EpochSetup event is expected, but we emit EpochCommit here resulting in a chunk fault
+	epochCommitServiceEvent, err := convert.ServiceEvent(testChain, epochCommitEvent)
+	require.NoError(s.T(), err)
+
+	s.snapshots[script] = &snapshot.ExecutionSnapshot{}
+	s.outputs[script] = fvm.ProcedureOutput{
+		ComputationUsed:        computationUsed,
+		ConvertedServiceEvents: flow.ServiceEventList{*epochCommitServiceEvent},
+		Events:                 meta.ChunkEvents[:3],
+	}
+
+	_, err = s.verifier.Verify(vch)
+
+	assert.Error(s.T(), err)
+	assert.True(s.T(), chunksmodels.IsChunkFaultError(err))
+	assert.IsType(s.T(), &chunksmodels.CFInvalidServiceEventsEmitted{}, err)
+}
+
+// Tests that service events are checked, when they appear outside the system chunk.
+// NOTE: this test case relies on the ordering of transactions in generateCollection.
+func (s *ChunkVerifierTestSuite) TestServiceEventsAreChecked_NonSystemChunk() {
+	script := "service event in non-system chunk"
+	meta := s.GetTestSetup(s.T(), script, false, true)
+	vch := meta.RefreshChunkData(s.T())
+
+	// setup the verifier output to include the correct data for the service events
+	output := generateDefaultOutput()
+	output.ConvertedServiceEvents = meta.ServiceEvents
+	output.Events = meta.ChunkEvents[:3] // 2 default events + 1 service event
+	s.outputs[script] = output
+
+	spockSecret, err := s.verifier.Verify(vch)
+	assert.NoError(s.T(), err)
+	assert.NotNil(s.T(), spockSecret)
+}
```
> For me, it would really help if TestServiceEventsMismatch_NonSystemChunk mirrored TestServiceEventsAreChecked_NonSystemChunk with as little changes as possible.

They are different because the existing testing infrastructure has very different code-paths for the system chunk and other chunks. Unfortunately I don't think it is feasible to make them more similar without a larger refactor of this test file.

jordanschalm commented on module/chunks/chunkVerifier_test.go:
```diff
@@ -282,9 +289,54 @@ func (s *ChunkVerifierTestSuite) TestServiceEventsAreChecked() {
+	epochCommitServiceEvent, err := convert.ServiceEvent(testChain, epochCommitEvent)
+	require.NoError(s.T(), err)
+
+	s.snapshots[script] = &snapshot.ExecutionSnapshot{}
+	s.outputs[script] = fvm.ProcedureOutput{
+		ComputationUsed:        computationUsed,
+		ConvertedServiceEvents: flow.ServiceEventList{*epochCommitServiceEvent},
+		Events:                 meta.ChunkEvents[:3],
+	}
```
I think your comment got cut off. The portion of code you linked is constructing an expected output for a transaction in which a service event was emitted (outside the system chunk).
- Line 328 is assigning the default service events for the output for a non-system-chunk transaction, as the expected output
- Line 329 is pulling out the 3 events associated with the transaction, and adding it to the expected output
- Line 330 is inserting the expected output into the map, so the verifier will consider this the

jordanschalm commented on module/chunks/chunkVerifier_test.go:
```diff
@@ -282,9 +289,54 @@ func (s *ChunkVerifierTestSuite) TestServiceEventsAreChecked() {
+	// setup the verifier output to include the correct data for the service events
+	output := generateDefaultOutput()
+	output.ConvertedServiceEvents = meta.ServiceEvents
+	output.Events = meta.ChunkEvents[:3] // 2 default events + 1 service event
```
The output is the expected output for _one transaction_. The test framework adds 2 events (contents of `eventsList`) as the expected output for every transaction by default. If specified (new option in this PR), it will additionally add 1 service event (3 total).
